### DISCUSSION TITLE
Union 2 mesh ?

### DISCUSSION THREAD

**Discussion by Tchiimy** [kinnala/scikit-fem#707] [Q&A]
Hi, 

For my project I would like to use scikitfem on a rpi4 to do FEM of a cantilever on which I'll add a mass to tune its frequency.

To represent what I would like to do here some examples from COMSOL : 

**Beam only :**
![image](https://user-images.githubusercontent.com/75799578/120444777-3eb5c300-c388-11eb-941a-29c0852b1709.png)

**Beam + added_mass :** 

![image](https://user-images.githubusercontent.com/75799578/120444882-5ab96480-c388-11eb-8685-8beab09a8f34.png)

I've done : 

```
from skfem.visuals.matplotlib import draw, show, plot
from skfem.models.elasticity import linear_elasticity, lame_parameters
from skfem.helpers import dot
from skfem import *
import numpy as np

print("import")

# Beam parameter
print("parameters")
w = 0.3
h = 0.05
t = 0.03

# added_mass parameter
w2 = 0.03
h2 = 0.03
t2 = 0.03

x0 = 0.0125
y0 = 0.25
z0 = 0.03

# physics parameters
rho = 1120
E = 2.174e9
nu = 0.35


# Beam Mesh
print("geom_beam")
beam = np.array([[0, 0, 0],
                 [0., 0., t],
                 [0., w, 0.],
                 [h, 0., 0.],
                 [0., w, t],
                 [h, 0, t],
                 [h, w, 0.],
                 [h, w, t]], dtype=np.float64).T

m1 = MeshTet(beam).refined(3)
print(m1)
draw(m1)
show()

# added mass mesh
print("geom_addedmass")


added_mass = np.array([[x0, y0, z0],
                       [x0, y0, z0 + t2],
                       [x0, y0 + w2, z0],
                       [x0 + h2, y0, z0],
                       [x0, y0 + w2, z0 + t2],
                       [x0 + h2, y0, z0 + t2],
                       [x0 + h2, y0 + w2, z0],
                       [x0 + h2, y0 + w2, z0 + t2]], dtype=np.float64).T

print(beam + added_mass)

m2 = MeshTet(added_mass).refined(3)

print(m2)
draw(m2)
show()

m = m1 + m2
draw(m)
show()


# Define FEM elements
print("FEM")
e1 = ElementTetP2()
e = ElementVectorH1(e1)
ib = InteriorBasis(m, e)


# Boundaries ?
print("Boundaries")

test = m.facets_satisfying(lambda x: x[0] == 0)
print(test)
test2 = m.facets_satisfying(lambda x: x[1] == 0)
print(test2)

dofs = {
    'left': ib.get_dofs(lambda x: x[0] == 0.0),
    'left': ib.get_dofs(lambda x: x[1] == 0.0),
}


# Modal Freqs
print("bilinear form definition")


@ BilinearForm
def mass(u, v, w):
    return dot(rho * u, v)


# K, M matrice
print("K,M matrice")

K = asm(linear_elasticity(*lame_parameters(E, nu)), ib)
M = asm(mass, ib)

u = ib.zeros()
I = ib.complement_dofs(dofs)

L, x = solve(*condense(K, M, I=I),
             solver=solver_eigen_scipy_sym(k=6, sigma=0.0))

y = x[:, 1]

Freqs = np.sqrt(L) / (2 * np.pi)
print(Freqs)


if __name__ == "__main__":
    sf = 0.01
    draw(MeshTet(np.array(m.p + sf * x[ib.nodal_dofs, 1]), m.t), colobar=True)
    show()
```

And I got  : 

![image](https://user-images.githubusercontent.com/75799578/120446560-f7303680-c389-11eb-9958-62b3305989e9.png)

But then when I try to do the solver, I get this : 

![image](https://user-images.githubusercontent.com/75799578/120446760-2b0b5c00-c38a-11eb-843f-ed6b00140205.png)

I don't think that's the good way to do an "union" between 2 3D mesh. I've tried to look in the documentation or the examples but I found nothing.

Any suggestions ?

**Comment by kinnala** [comment]
If you sum meshes, i.e. `m = m1 + m2`, you must verify that the nodes match. Otherwise the result is not a valid mesh.

The geometry looks nontrivial so I suggest to use an external mesh generator. For example:
- Export mesh from COMSOL to `mphtxt`, convert to `.msh` e.g. using FEconv, load to scikit-fem via `MeshTet.load("beam.msh")`
- Generate the mesh using Gmsh, there are several examples under `docs/examples/`

If you want to do it entirely inside scikit-fem, then you most likely have to use `MeshHex.init_tensor` to create two meshes where the nodes match and then perform `m = m1 + m2`. Finally, you can split it to tetrahedra using `MeshHex.to_meshtet`.

**Comment by Tchiimy** [comment]
To update you : 

I've succeeded to create a 3D mesh with an union between the beam and added_mass with [this](https://jsdokken.com/converted_files/tutorial_pygmsh.html) and it seems to work fine when I import it on scikit-fem.

By curiosity, I'll try to do the entirely inside scikit-fem solution, if I understood i need to create 2 mesh with MeshHex.init_tensor with common nodes for the union part, then I addition them and I do the meshing with MeshHex.to_meshtet.

**Comment by kinnala** [comment]
> By curiosity, I'll try to do the entirely inside scikit-fem solution, if I understood i need to create 2 mesh with MeshHex.init_tensor with common nodes for the union part, then I addition them and I do the meshing with MeshHex.to_meshtet.

Yes!
